// umi-build-dev/lib/plugins/commands/generate/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

var _path = require("path");

var _assert = _interopRequireDefault(require("assert"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _default(api) {
  const generators = api.service.generators;

  function generate(args = {}) {
    const name = args._[0];
    (0, _assert.default)(name, `Invalid arguments, try to use *umi generate name {args}*`);
    (0, _assert.default)(generators[name], `Generator ${name} not found`);
    const _generators$name = generators[name],
          Generator = _generators$name.Generator,
          resolved = _generators$name.resolved;
    const generator = new Generator(process.argv.slice(4), {
      env: {
        cwd: api.cwd
      },
      resolved: resolved || __dirname
    });
    generator.run(() => {
      console.log('Done');
    });
  }

  api.registerCommand('g', {
    description: 'generate code snippets quickly (alias for generate)'
  }, generate);
  api.registerCommand('generate', {
    description: 'generate code snippets quickly'
  }, generate);
  api.registerGenerator('page', {
    Generator: require('./page').default(api),
    resolved: (0, _path.join)(__dirname, './page')
  });
}
